repo/modules.go

```diff
@@ -52,6 +52,7 @@ func importRepoRulesModules(filename string, _ *RemoteCache) (repos []Repo, err
 		}
 	}
 	pathToModule := map[string]*module{}
+	replacePathToModule := map[string]*module{}
```

Comment:
We seem to be duplicating a lot of logic below. At the end of this loop, let's make `pathToModule` map importable module paths to `*module`. So for example, say we have

```
module M

require (
  A v1.0.0
  B v1.0.0
)

replace A v1.0.0 => B v1.0.0
```

`pathToModule` should map `A` to a module with `m.Path == "A"` and `m.Replace.Path == "B"`. `pathToModule` should not contain a separate entry for `B`.

If we see `B` first in the list, we should delete it from `pathToModule` when we see `A`. If we see `A` first, we should avoid adding `B`.

Reply:
The example I mention in #521 (and the one you do above) will not be built by go: https://go-review.googlesource.com/c/go/+/126156/. 

When you call `go mod why` you get
```
go: a@v1.0.0 used for two different module paths (a and b)
```
With this in mind, I think both modules are equally valid/invalid, so it doesn't make sense to only generate a rule for the module with the replace directive. We should either generate rules for both modules or print a warning and not generate rules for either.

Also, it is a valid case when a single module has two different paths, as long as the paths don't have the same version:
```
module M

require (
  A v1.0.0
  B v1.0.0
)

replace A v1.0.0 => B v1.2.0
```

`go list -m all` returns
```
A v1.0.0 => B v1.2.0
B v1.0.0
```
And we will want two `go_repository` rules:
```
go_repository(
  importpath = "B",
  version = "1.0.0",
  sum = <sum_from_B@1.0.0>,
)
go_repository(
  importpath = "A",
  version = "1.2.0",
  replace = "B",
  sum = <sum_from_B@1.2.0>,
)
```

In this case, we will need both modules. I have thought of a way to consolidate the logic a little bit (always use `mod.Path` as `pathToModule`'s index, create an importpath=>sum map of the `go.sum` file and then iterate through `pathToModule` searching for matching sums using `mod.Replace.Path` if there is a replace directive.)

Reply:
> With this in mind, I think both modules are equally valid/invalid, so it doesn't make sense to only generate a rule for the module with the replace directive. We should either generate rules for both modules or print a warning and not generate rules for either.

You're right, I just tried this with my fork of bazel-gazelle as a concrete example, and `go list -m -json all` reports an error. If that happens, we shouldn't generate any rules.

> Also, it is a valid case when a single module has two different paths, as long as the paths don't have the same version

Agreed, we should generate two `go_repository` rules for this. I mainly want to avoid doing anything different than what the Go command does. We shouldn't create two `go_repository` rules with different names for the same bits, but it should be okay if they're at different versions.

Reply:
> You're right, I just tried this with my fork of bazel-gazelle as a concrete example, and go list -m -json all reports an error. If that happens, we shouldn't generate any rules.

Interesting, I tried reproing with go 1.12.5 and did not get an error when calling `go list -m -json all`, only with `go mod why`... For older versions of go where `go list` does not throw an error in this scenario, should Gazelle treat it as an error or generate both of the rules?  

Going off of your desire to not sway from the Go command, I guess we should ignore the error. The user will get an error as soon as they try to build, and will then proceed to fix their `go.mod` file and re-run `update-repos`.

Reply:
I was seeing this error in 1.12.5, too. Here's my go.mod file:

```
module m

require (
	github.com/bazelbuild/bazel-gazelle v0.0.0-20190513151847-b16e182bf090 // indirect
	github.com/jayconrod/bazel-gazelle v0.0.0-20190513151847-b16e182bf090
)

replace github.com/bazelbuild/bazel-gazelle => github.com/jayconrod/bazel-gazelle v0.0.0-20190513151847-b16e182bf090
```

In any case though, I think treating `go list -m -json all` as the source of truth is our best option here. The semantics and errors will change over time, and we'll need to adjust.